### PR TITLE
Initial implementation of acceleration structures

### PR DESCRIPTION
**Pull request by HansKristian-Work**
Goal here is to implement enough so that test can pass.

Added some extra test for post-build info, and it passes on NV on the tests which are not marked TODO. There are some queries that we cannot support without lots of heroics, but they seem somewhat useless for applications to query, so I'd like to defer those until we really need to care.

### CODE REVIEW

File: libs/vkd3d/device.c [hunk 1]
```diff
@@ -4223,7 +4223,43 @@ static void STDMETHODCALLTYPE d3d12_device_GetRaytracingAccelerationStructurePre
...
+    if (!device->device_info.acceleration_structure_features.accelerationStructure)
+    {
+        ERR("Acceleration structure is not supported. Calling this is invalid.\n");
+        memset(info, 0, sizeof(*info));
+        return;
+    }
+
+    if (!vkd3d_acceleration_structure_convert_inputs(&build_info, desc))
+    {
+        ERR("Failed to convert inputs..\n");
```
Comment (Guy1524) [inline libs/vkd3d/device.c:4250]: frog

File: libs/vkd3d/acceleration_structure.c [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    if (desc->Flags & D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAG_ALLOW_UPDATE)
+        build_info->flags |= VK_BUILD_ACCELERATION_STRUCTURE_ALLOW_UPDATE_BIT_KHR;
+    if (desc->Flags & D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAG_MINIMIZE_MEMORY)
+        build_info->flags |= VK_BUILD_ACCELERATION_STRUCTURE_LOW_MEMORY_BIT_KHR;
+    if (desc->Flags & D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAG_PREFER_FAST_BUILD)
+        build_info->flags |= VK_BUILD_ACCELERATION_STRUCTURE_PREFER_FAST_BUILD_BIT_KHR;
+    if (desc->Flags & D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAG_PREFER_FAST_TRACE)
+        build_info->flags |= VK_BUILD_ACCELERATION_STRUCTURE_PREFER_FAST_TRACE_BIT_KHR;
+
+    if (desc->Flags & D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAG_PERFORM_UPDATE)
```
Comment (Joshua-Ashton) [inline libs/vkd3d/acceleration_structure.c:65]: Maybe have helper funcs for converting these?

File: libs/vkd3d/device.c (hunk 1 above)
Comment (HansKristian-Work) [reply libs/vkd3d/device.c]: Fixed.

File: libs/vkd3d/acceleration_structure.c (hunk 2 above)
Comment (HansKristian-Work) [reply libs/vkd3d/acceleration_structure.c:75]: Fixed.

File: libs/vkd3d/acceleration_structure.c [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    VK_CALL(vkCmdWriteAccelerationStructuresPropertiesKHR(list->vk_command_buffer,
+            1, &vk_acceleration_structure, vk_query_type, vk_query_pool, vk_query_index));
+    VK_CALL(vkCmdCopyQueryPoolResults(list->vk_command_buffer,
+            vk_query_pool, vk_query_index, 1,
+            vk_buffer, offset, stride,
+            VK_QUERY_RESULT_64_BIT | VK_QUERY_RESULT_WAIT_BIT));
+
+    if (desc->InfoType == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_POSTBUILD_INFO_SERIALIZATION)
+    {
+        /* We'll need some way to store these values for later use and copy them here instead. */
```
Comment (doitsujin) [inline libs/vkd3d/acceleration_structure.c:267]: Is this a TODO? If so, probably should be marked as such.

**Review by doitsujin — COMMENTED** [review]
I suppose the RTAS code itself is fine if it passes tests+validation, just a nit otherwise.

File: libs/vkd3d/vkd3d_private.h [hunk 4]
```diff
@@ -2671,6 +2700,35 @@ struct vkd3d_view_key
 struct vkd3d_view *vkd3d_view_map_create_view(struct vkd3d_view_map *view_map,
         struct d3d12_device *device, const struct vkd3d_view_key *key);
 
+/* Acceleration structure helpers. */
+struct vkd3d_acceleration_structure_build_info
+{
+#define VKD3D_BUILD_INFO_STACK_COUNT 16
```
[the remark this answers is not in the document]
Comment (HansKristian-Work) [reply libs/vkd3d/vkd3d_private.h:2708]: Fixed.

File: libs/vkd3d/command.c [hunk 5]
```diff
@@ -6073,12 +6087,35 @@ static void STDMETHODCALLTYPE d3d12_command_list_ResourceBarrier(d3d12_command_l
...
+                        const VkPipelineStageFlags rt_stages = VK_PIPELINE_STAGE_RAY_TRACING_SHADER_BIT_KHR |
+                                VK_PIPELINE_STAGE_ACCELERATION_STRUCTURE_BUILD_BIT_KHR;
+                        const VkAccessFlags rt_access = VK_ACCESS_ACCELERATION_STRUCTURE_WRITE_BIT_KHR |
+                                VK_ACCESS_ACCELERATION_STRUCTURE_READ_BIT_KHR;
+                        src_stage_mask |= rt_stages;
+                        dst_stage_mask |= rt_stages;
+                        vk_memory_barrier.srcAccessMask |= rt_access;
+                        vk_memory_barrier.dstAccessMask |= rt_access;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (HansKristian-Work) [reply libs/vkd3d/command.c]: Should be fixed.

File: libs/vkd3d/acceleration_structure.c (hunk 3 above)
Comment (HansKristian-Work) [reply libs/vkd3d/acceleration_structure.c]: Fixed.

File: libs/vkd3d/vkd3d_private.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (HansKristian-Work) [reply libs/vkd3d/vkd3d_private.h:2708]: Actually, fixed in the commit itself instead.

**Review by doitsujin — APPROVED** [review]

**Pull request merged**